File: ISchemm.UTF32.Arrays/Char32Array.cs
using System;
using System.Runtime.InteropServices;

namespace ISchemm.UTF32.Arrays
{
    public static class Char32Array
    {
        public unsafe static Char32[] FromByteArray(byte[] array)
        {
            if (array.Length % sizeof(Char32) != 0)
                throw new FormatException($"Length of array must be divisible by {sizeof(Char32)}");

            Char32[] destArray = new Char32[array.Length / sizeof(Char32)];
            fixed (Char32* dest = destArray) {
                Marshal.Copy(array, 0, (IntPtr)dest, array.Length);
            }
            return destArray;
        }

        public static Char32[] FromString(string str)
        {
            return FromByteArray(Char32.Encoding.GetBytes(str));
        }

        public unsafe static byte[] GetByteArray(this Char32[] array)
        {
            byte[] destArray = new byte[array.Length * sizeof(Char32)];
            fixed (Char32* src = array)
            {
                Marshal.Copy(new IntPtr(src), destArray, 0, destArray.Length);
            }
            return destArray;
        }

        public static string GetString(this Char32[] array)
        {
            return Char32.Encoding.GetString(GetByteArray(array));
        }
    }
}

File: ISchemm.UTF32.Arrays/Char32Array.fs
#nowarn "9"

namespace ISchemm.UTF32.Arrays

open System
open System.Runtime.InteropServices
open Microsoft.FSharp.NativeInterop
open ISchemm.UTF32

module Char32Array =
    let FromByteArray (bArr: byte[]): Char32[] =
        if bArr.Length % sizeof<Char32> <> 0 then
            raise (FormatException $"Length of array must be divisible by {sizeof<Char32>}")

        let cArr = Array.zeroCreate<Char32> (bArr.Length / sizeof<Char32>)
        use cPtr = fixed cArr in Marshal.Copy (bArr, 0, NativePtr.toNativeInt cPtr, bArr.Length)
        cArr

    let FromString (str: string): Char32[] =
        str
        |> Char32.Encoding.GetBytes
        |> FromByteArray

    let GetByteArray (cArr: Char32[]): byte[] =
        let bArr = Array.zeroCreate<byte> (cArr.Length * sizeof<Char32>)
        use cPtr = fixed cArr in Marshal.Copy (NativePtr.toNativeInt cPtr, bArr, 0, bArr.Length)
        bArr

    let GetString (array: Char32[]): string =
        array
        |> GetByteArray
        |> Char32.Encoding.GetString
